fdbserver/CommitProxyServer.actor.cpp

```diff
@@ -1063,6 +1071,95 @@ void assertResolutionStateMutationsSizeConsistent(const std::vector<ResolveTrans
 	}
 }
 
+// Return true if a single-key mutation is associated with a valid tenant id or a system key
+bool validTenantAccess(MutationRef m, std::map<int64_t, TenantName> const& tenantMap, Optional<int64_t>& tenantId) {
+	if (isSingleKeyMutation((MutationRef::Type)m.type)) {
+		tenantId = TenantAPI::extractTenantIdFromMutation(m);
+		return tenantMap.count(tenantId.get()) > 0;
+	}
+	return true;
+}
+
+inline bool tenantMapChanging(MutationRef const& mutation) {
+	const KeyRangeRef tenantMapRange = TenantMetadata::tenantMap().subspace;
+	if (isSingleKeyMutation((MutationRef::Type)mutation.type) && mutation.param1.startsWith(tenantMapRange.begin)) {
+		return true;
+	} else if (mutation.type == MutationRef::ClearRange &&
+	           tenantMapRange.intersects(KeyRangeRef(mutation.param1, mutation.param2))) {
+		return true;
+	}
+	return false;
+}
+
+// Return success and properly split clear range mutations if all tenant check pass. Otherwise, return corresponding
+// error
+Error validateAndProcessTenantAccess(VectorRef<MutationRef> mutations,
+                                     ProxyCommitData* const pProxyCommitData,
+                                     Optional<UID> debugId = Optional<UID>(),
+                                     const char* context = "") {
+	bool changeTenant = false;
+	bool writeNormalKey = false;
+
+	for (auto& mutation : mutations) {
+		Optional<int64_t> tenantId;
+		bool validAccess = true;
+		changeTenant = changeTenant || tenantMapChanging(mutation);
+
+		if (mutation.type == MutationRef::ClearRange) {
+			auto beginTenantId = TenantAPI::extractTenantIdFromKeyRef(mutation.param1);
+			auto endTenantId = TenantAPI::extractTenantIdFromKeyRef(mutation.param2);
+			CODE_PROBE(beginTenantId != endTenantId, "Clear Range raw access or cross multiple tenants");
+			// TODO: splitting clear range
+		} else if (!isSystemKey(mutation.param1)) {
+			validAccess = validTenantAccess(mutation, pProxyCommitData->tenantMap, tenantId);
+			writeNormalKey = true;
+		}
+
+		if (debugId.present()) {
+			TraceEvent(SevDebug, "ValidateAndProcessTenantAccess", pProxyCommitData->dbgid)
+			    .detail("Context", context)
+			    .detail("TxnId", debugId.get())
+			    .detail("Version", pProxyCommitData->version.get())
+			    .detail("ChangeTenant", changeTenant)
+			    .detail("WriteNormalKey", writeNormalKey)
+			    .detail("TenantId", tenantId)
+			    .detail("ValidAccess", validAccess)
+			    .detail("MutationType", getTypeString(mutation.type))
+			    .detail("Mutation", mutation.param1);
+		}
+
+		if (!validAccess) {
+			TraceEvent(SevWarn, "TenantAccessCheck_IllegalTenantWrite", pProxyCommitData->dbgid)
```

Comment:
Underscores should generally be avoided in trace events if possible. For this, I'd recommend:

```suggestion
			TraceEvent(SevWarn, "IllegalTenantAccess", pProxyCommitData->dbgid)
```

I'm also not sure what our policy is on logging keys, maybe @sfc-gh-ahusain could comment?